### PR TITLE
State diff

### PR DESCRIPTION
**Pull request by winsvega**
add --statediff option when filling or running tests

### CODE REVIEW

File: libethereum/State.cpp [hunk 1]
```diff
@@ -541,6 +541,7 @@ std::pair<ExecutionResult, TransactionReceipt> State::execute(EnvInfo const& _en
 	u256 const startGasUsed = _envInfo.gasUsed();
 	executeTransaction(e, _t, onOp);
 
+	res.changeLog = m_changeLog;
```
Comment (gumb0) [inline libethereum/State.cpp]: I don'like that this way we will have this in `ExecutionResult` (paying with additional memory consumption) even when we don't need it (we need it only for tests and maybe vmtraces)

@chfast what do you think?

Comment (winsvega) [inline libethereum/State.cpp]: could be under cmake flag --VMTRACE

File: libdevcore/CommonData.h [hunk 2]
```diff
@@ -172,9 +179,9 @@ inline std::string toHex(u256 val, HexPrefix prefix = HexPrefix::DontAdd)
 	return (prefix == HexPrefix::Add) ? "0x" + str : str;
 }
 
-inline std::string toCompactHex(u256 val, HexPrefix prefix = HexPrefix::DontAdd, unsigned _min = 0)
+inline std::string toCompactHex(u256 val, HexPrefix prefix = HexPrefix::DontAdd, PrintZero _min = PrintZero::DontPrint)
```
Comment (gumb0) [inline libdevcore/CommonData.h]: So many places where we pass `PrintZero::Print` here, is it ever used really with `PrintZero::DontPrint` ?

File: test/tools/libtesteth/ImportTest.cpp [hunk 3]
```diff
@@ -629,6 +654,10 @@ int ImportTest::exportTest(bytes const& _output)
 				if (it != std::end(stateIndexesToPrint))
 					obj2["postState"] = fillJsonWithState(m_transactions[i].postState);
 			}
+
+			if (Options::get().statediff)
```
Comment (gumb0) [inline test/tools/libtesteth/ImportTest.cpp]: Does this go into generated json files?

File: test/tools/libtesteth/TestHelper.cpp [hunk 4]
```diff
@@ -124,19 +125,126 @@ eth::Network stringToNetId(string const& _netname)
...
 
+json_spirit::mObject fillJsonWithStateChange(State const& _stateOrig, eth::State const& _statePost, std::vector<dev::eth::detail::Change> const& _changeLog)
+{
+	json_spirit::mObject oState;
+	typedef dev::eth::detail::Change::Kind changeKind;
+	typedef std::vector<dev::eth::detail::Change> logVector;
+	std::vector<dev::eth::detail::Change> changeLog = _changeLog;
+
+	std::ostringstream log;
+	while(changeLog.size())
```
Comment (gumb0) [inline test/tools/libtesteth/TestHelper.cpp]: Why do you need to iterate in this weird way instead of just `for (auto& change : changeLog)` ?

**Comment by gumb0** [comment]
It probably would be nice to have this feature not only for tests, but also for regular `eth` operation, like we have vmtrace for both... Would it be difficult to output state diff to log in `eth` when, say, VMTRACE is enabled ?

File: libethereum/State.cpp (hunk 1 above)
Comment (chfast) [inline libethereum/State.cpp]: I don't like it either. This must go to `dev::test` namespace unless proven it is not possible to be done this way.

And big no to build flags. Let's find a way to do it nicely without affecting not-test related code.

File: libdevcore/CommonData.h (hunk 2 above)
Comment (chfast) [inline libdevcore/CommonData.h]: This looks bad. Can you describe what the use cases for this function are? If tests needs something different maybe it will be better for `dev::test` to have a separated function for this case.

**Comment by chfast** [comment]
Ok, let's take is slowly.
Can we separate the changes related to `toCompactHex` from the "state diff" one first?

File: libethereum/State.cpp (hunk 1 above)
Comment (gumb0) [inline libethereum/State.cpp]: I had in mind adding another callback function for collecting state diffs as `State::execute` parameter.

Comment (chfast) [inline libethereum/State.cpp]: Is the "state diff" the same as "state journal" or "state changelog"? If yes, let's stick to a single name. We can add a function to export the changelog next to the functions like `savepoint()` or `revert()`.

Comment (gumb0) [inline libethereum/State.cpp]: I think it's essentially the same, let's call it change log.

We need it after transaction execution (`executeTransaction` call on line 542), but before `commit` which clears changelog.

The getter function would work if we save the copy before commit, or don't clear it.

Comment (winsvega) [inline libethereum/State.cpp]: yes, getter function would be nice. 
state.executeTransaction()
state.getChangelog()

File: libdevcore/CommonData.h (hunk 2 above)
Comment (winsvega) [inline libdevcore/CommonData.h]: there is a person who constantly asking me to change hex format in our tests. I've changed it here to PrintZero::DontPrint because that is what the meaning of that "1" was in the code. 
but the changes are not final.

File: test/tools/libtesteth/ImportTest.cpp (hunk 3 above)
Comment (winsvega) [inline test/tools/libtesteth/ImportTest.cpp]: when using with --filltests yes.

File: test/tools/libtesteth/TestHelper.cpp (hunk 4 above)
Comment (winsvega) [inline test/tools/libtesteth/TestHelper.cpp]: I use this because of the changeLog.erase

File: test/tools/libtesteth/ImportTest.cpp (hunk 3 above)
Comment (gumb0) [inline test/tools/libtesteth/ImportTest.cpp]: Do we really need to write it into files and incerase their size and complexity? It's only for human readers, right?

Maybe just leave it like a log option for `testeth` run ?

File: test/tools/libtesteth/TestHelper.cpp (hunk 4 above)
Comment (gumb0) [inline test/tools/libtesteth/TestHelper.cpp]: Why do you need to erase while iterating over it?

Comment (gumb0) [inline test/tools/libtesteth/TestHelper.cpp]: If you need group changes by the addresses, you could sort `changeLog` vector by `address` field, then do all the processing in one iteration over the vector

**Comment by winsvega** [comment]
refactored the log function. 
I've removed changeLog clean function from commit in state. and moved it to the begining  of execute.
will see how tests work. if there is an issue we could dublicate m_changeLog variable in State just for  getChangeLog() method

File: test/tools/libtesteth/FillJsonFunctions.cpp [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+
+	std::ostringstream log;
+	json_spirit::mObject o;
+
+	json_spirit::mArray agregatedBalance;
+	json_spirit::mArray agregatedStorage;
+	changeStruct popChange = changeLog.at(0);
+	std::map<string, ostringstream> logInfo;
+	u256 tmpBalance = _stateOrig.balance(popChange.address);
+	for (size_t i = 0; i < changeLog.size(); i++)
```
Comment (gumb0) [inline test/tools/libtesteth/FillJsonFunctions.cpp]: This loop probably works correctly, but I think expressing it something like this would be clearer:
```
size_t i = 0;
while (i < changeLog.size())
{
	Addres currentAddress = changeLog[i].address;

	string before;
	string after;
	json_spirit::mArray record;
	json_spirit::mArray agregatedBalance;
	json_spirit::mArray agregatedStorage;
	std::map<string, ostringstream> logInfo;

	while (i < changeLog.size() && changeLog[i].address == currentAddress)
	{
		// handle changeLog[i]

		++i;
	}
}
```

**Comment by pirapira** [comment]
I think ChangLog.md can talk about the new option.

File: test/tools/libtesteth/FillJsonFunctions.cpp [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+	{
+		if (_map.size() && _map.find(a.first) == _map.end())
+			continue;
+
+		json_spirit::mObject o;
+		if (mapEmpty || _map.at(a.first).hasBalance())
+			o["balance"] = toCompactHex(_state.balance(a.first), HexPrefix::Add, 1);
+		if (mapEmpty || _map.at(a.first).hasNonce())
+			o["nonce"] = toCompactHex(_state.getNonce(a.first), HexPrefix::Add, 1);
+		{
```
Comment (pirapira) [inline test/tools/libtesteth/FillJsonFunctions.cpp]: Why is a `{` here?  The `if` statement has ended in the last line.

File: test/tools/libtesteth/ImportTest.cpp [hunk 7]
```diff
@@ -637,6 +667,8 @@ int ImportTest::exportTest(bytes const& _output)
 			obj[it->first] = it->second;
 
 		m_testObject["post"] = obj;
+
+
```
Comment (pirapira) [inline test/tools/libtesteth/ImportTest.cpp]: Spurious lines?

File: libethereum/StateDetail.h [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+
+	cpp-ethereum is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with cpp-ethereum.  If not, see <http://www.gnu.org/licenses/>.
+*/
+/** @file StateDetail.h
```
Comment (chfast) [inline libethereum/StateDetail.h]: Why did you move it to another file?

**Comment by winsvega** [comment]
the first solution was to put changeLog into execution results. that required separation of StateDetail from State

File: libethereum/StateDetail.h (hunk 8 above)
Comment (chfast) [inline libethereum/StateDetail.h]: You have to move it back.

File: test/tools/libtesteth/FillJsonFunctions.cpp [hunk 9]
```diff
@@ -0,0 +1,227 @@
...
+
+	cpp-ethereum is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with cpp-ethereum.  If not, see <http://www.gnu.org/licenses/>.
+*/
+/** @file
```
Comment (chfast) [inline test/tools/libtesteth/FillJsonFunctions.cpp]: What is the point of not having these function in TestHelper.cpp?

**Comment by winsvega** [comment]
to split the large file with lots of functions into smaller ones

**Comment by chfast** [comment]
This just makes it compile twice as long as before.

**Comment by winsvega** [comment]
I stick to that guy's opinion:
"f I find that an implementation file is thousands of lines long, that's usually a sign that there's too much there and I need to break it up."
https://stackoverflow.com/questions/531133/should-i-put-many-functions-into-one-file-or-more-or-less-one-function-per-fi

File: test/tools/libtesteth/FillJsonFunctions.cpp (hunk 5 above)
Comment (chfast) [inline test/tools/libtesteth/FillJsonFunctions.cpp]: @winsvega ?

File: libethereum/State.cpp [hunk 10]
```diff
@@ -224,7 +225,6 @@ void State::commit(CommitBehaviour _commitBehaviour)
 	if (_commitBehaviour == CommitBehaviour::RemoveEmptyAccounts)
 		removeEmptyAccounts();
 	m_touched += dev::eth::commit(m_cache, m_state);
-	m_changeLog.clear();
```
Comment (chfast) [inline libethereum/State.cpp]: `commit()` have to clear the changelog. See comments below.

File: test/tools/libtesteth/ImportTest.cpp (hunk 3 above)
Comment (chfast) [inline test/tools/libtesteth/ImportTest.cpp]: Answer for this?

File: test/tools/libtesteth/TestHelper.cpp [hunk 11]
```diff
@@ -477,7 +422,10 @@ void executeTests(const string& _name, const string& _testPathAppendix, const st
 	}
 	try
 	{
-		cnote << "TEST " << name << ":";
+		if ((Options::get().singleTest && Options::get().singleTestName == name)
+			||	!Options::get().singleTest)
```
Comment (chfast) [inline test/tools/libtesteth/TestHelper.cpp]: WTF is this tab in the middle?

**Comment by winsvega** [comment]
staediff also work when filling tests producing output to the test .json for debug

File: libethereum/State.cpp [hunk 12]
```diff
@@ -223,9 +224,12 @@ void State::commit(CommitBehaviour _commitBehaviour)
 {
 	if (_commitBehaviour == CommitBehaviour::RemoveEmptyAccounts)
 		removeEmptyAccounts();
+
 	m_touched += dev::eth::commit(m_cache, m_state);
-	m_changeLog.clear();
+	if (m_changeLog.size() != 0)
+		m_changeLogResult = m_changeLog;
```
Comment (chfast) [inline libethereum/State.cpp]: This is bad and inefficient (both fixable), but read below.

File: test/tools/libtesteth/ImportTest.cpp [hunk 13]
```diff
@@ -93,6 +93,7 @@ bytes ImportTest::executeTest()
 				eth::Network network = networks[j];
 				std::pair<eth::State, ImportTest::execOutput> out = executeTransaction(network, m_envInfo, m_statePre, m_transactions[i].transaction);
```
Comment (chfast) [inline test/tools/libtesteth/ImportTest.cpp]: You should try to improve `executeTransaction()` function first.
1. Do you need preState after this function. If not do not copy the State.
2. Can it also return changelog if you need it?
3. Use `std::tie()` to unpack tuple.

File: libethereum/State.cpp [hunk 14]
```diff
@@ -223,9 +223,10 @@ void State::commit(CommitBehaviour _commitBehaviour)
 {
 	if (_commitBehaviour == CommitBehaviour::RemoveEmptyAccounts)
 		removeEmptyAccounts();
+
 	m_touched += dev::eth::commit(m_cache, m_state);
-	m_changeLog.clear();
 	m_cache.clear();
+	m_changeLog.clear();
```
Comment (chfast) [inline libethereum/State.cpp]: Revert all changes here.

File: libethcore/ChainOperationParams.h [hunk 15]
```diff
@@ -15,8 +15,6 @@
 	along with cpp-ethereum.  If not, see <http://www.gnu.org/licenses/>.
 */
 /** @file ChainOperationsParams.h
- * @author Gav Wood <i@gavwood.com>
```
Comment (chfast) [inline libethcore/ChainOperationParams.h]: Do not do this unrelated changes.

File: libethereum/State.cpp [hunk 16]
```diff
@@ -541,12 +542,18 @@ std::pair<ExecutionResult, TransactionReceipt> State::execute(EnvInfo const& _en
...
-		m_cache.clear();
-	else
+	bool removeEmptyAccounts;
+	switch (_p)
 	{
-		bool removeEmptyAccounts = _envInfo.number() >= _sealEngine.chainParams().u256Param("EIP158ForkBlock");
-		commit(removeEmptyAccounts ? State::CommitBehaviour::RemoveEmptyAccounts : State::CommitBehaviour::KeepEmptyAccounts);
+		case Permanence::Reverted:
+			m_cache.clear();
+		break;
```
Comment (chfast) [inline libethereum/State.cpp]: `break` must be indented.

File: libethereum/State.cpp [hunk 17]
```diff
@@ -541,12 +542,18 @@ std::pair<ExecutionResult, TransactionReceipt> State::execute(EnvInfo const& _en
...
+		case Permanence::Committed:
+			removeEmptyAccounts = _envInfo.number() >= _sealEngine.chainParams().u256Param("EIP158ForkBlock");
+			commit(removeEmptyAccounts ? State::CommitBehaviour::RemoveEmptyAccounts : State::CommitBehaviour::KeepEmptyAccounts);
+		break;
+		default:
```
Comment (chfast) [inline libethereum/State.cpp]: Use `Unfinished` instead of `default`. Compilers will be satisfied by having cases for all options.

File: libethereum/State.h [hunk 18]
```diff
@@ -15,8 +15,6 @@
 	along with cpp-ethereum.  If not, see <http://www.gnu.org/licenses/>.
 */
 /** @file State.h
- * @author Gav Wood <i@gavwood.com>
```
Comment (chfast) [inline libethereum/State.h]: Unrelated change.

File: libethereum/State.h [hunk 19]
```diff
@@ -82,7 +80,8 @@ enum class BaseState
 enum class Permanence
 {
 	Reverted,
-	Committed
+	Committed,
+	Unfinished
```
Comment (chfast) [inline libethereum/State.h]: Documentation missing.

File: libethereum/State.h [hunk 20]
```diff
@@ -142,8 +141,7 @@ struct Change
 	{}
 };
 
-}
-
+} //detail
```
Comment (chfast) [inline libethereum/State.h]: Ugly formatting. Put 2 spaces before `//`, and space after `//`. Also better comment: `// namespace detail`.

File: libethereum/State.h [hunk 21]
```diff
@@ -336,7 +336,7 @@ class State
 	u256 m_accountStartNonce;
 
 	friend std::ostream& operator<<(std::ostream& _out, State const& _s);
-	std::vector<detail::Change> m_changeLog;
+	std::vector<detail::Change> m_changeLog;	///< Working changeLog
```
Comment (chfast) [inline libethereum/State.h]: 1. Working means the one that is not broken. You mean current, or change log of currently executing transaction?
2. "changeLog" is not an English word. You mean "change log"?
3. Missing full stop.

File: test/tools/libtesteth/ImportTest.cpp [hunk 22]
```diff
@@ -52,13 +53,14 @@ ImportTest::ImportTest(json_spirit::mObject& _o, bool isFiller, testType testTem
 
 bytes ImportTest::executeTest()
 {
+	using results = std::tuple<eth::State, ImportTest::ExecOutput, ImportTest::ChangeLog>;
```
Comment (chfast) [inline test/tools/libtesteth/ImportTest.cpp]: Type names must start with uppercase letter. Should be `Results`.

File: test/tools/libtesteth/ImportTest.cpp [hunk 23]
```diff
@@ -52,13 +53,14 @@ ImportTest::ImportTest(json_spirit::mObject& _o, bool isFiller, testType testTem
...
 	if (m_testType == testType::StateTests)
```
Comment (chfast) [inline test/tools/libtesteth/ImportTest.cpp]: Is this case needed any more?

File: test/tools/libtesteth/ImportTest.cpp [hunk 24]
```diff
@@ -91,8 +93,9 @@ bytes ImportTest::executeTest()
 					continue;
 
 				eth::Network network = networks[j];
-				std::pair<eth::State, ImportTest::execOutput> out = executeTransaction(network, m_envInfo, m_statePre, m_transactions[i].transaction);
-				m_transactions[i].postState = out.first;
+				results out = executeTransaction(network, m_envInfo, m_statePre, m_transactions[i].transaction);
```
Comment (chfast) [inline test/tools/libtesteth/ImportTest.cpp]: Use `std::tie()` to assign `postState` and `changeLog` and ignore tuple element of index 1.

File: test/tools/libtesteth/ImportTest.h [hunk 25]
```diff
@@ -54,15 +54,17 @@ class ImportTest
...
 
 	eth::State m_statePre;
 	eth::State m_statePost;
 	eth::LogEntries m_logs;
 	eth::LogEntries m_logsExpected;
 
 private:
-	typedef std::pair<eth::ExecutionResult, eth::TransactionReceipt> execOutput;
-	std::pair<eth::State, execOutput> executeTransaction(eth::Network const _sealEngineNetwork, eth::EnvInfo const& _env, eth::State const& _preState, eth::Transaction const& _tr);
+	typedef std::vector<dev::eth::detail::Change> ChangeLog;
```
Comment (chfast) [inline test/tools/libtesteth/ImportTest.h]: This must go to State.h. Also use it in State class declaration.

File: test/tools/libtesteth/ImportTest.h [hunk 26]
```diff
@@ -54,15 +54,17 @@ class ImportTest
...
+	typedef std::pair<eth::ExecutionResult, eth::TransactionReceipt> ExecOutput;
```
Comment (chfast) [inline test/tools/libtesteth/ImportTest.h]: Use `using` instead of `typedef` to make a type alias.

File: test/tools/libtesteth/TestHelper.cpp [hunk 27]
```diff
@@ -477,7 +422,9 @@ void executeTests(const string& _name, const string& _testPathAppendix, const st
 	}
 	try
 	{
-		cnote << "TEST " << name << ":";
+		if ((Options::get().singleTest && Options::get().singleTestName == name) || !Options::get().singleTest)
+			cnote << "TEST " << name << ":";
```
Comment (chfast) [inline test/tools/libtesteth/TestHelper.cpp]: Verbosity 2 outputs test names. Why do you need this?

Comment (winsvega) [inline test/tools/libtesteth/TestHelper.cpp]: so that when you use --statediff you would see only the name of the test you are running with --singletest option. 
other names make it hard to see the changeLog in console when there are many tests

File: libethereum/State.cpp [hunk 28]
```diff
@@ -541,12 +541,18 @@ std::pair<ExecutionResult, TransactionReceipt> State::execute(EnvInfo const& _en
 	u256 const startGasUsed = _envInfo.gasUsed();
 	executeTransaction(e, _t, onOp);
 
-	if (_p == Permanence::Reverted)
-		m_cache.clear();
-	else
+	bool removeEmptyAccounts;
```
Comment (chfast) [inline libethereum/State.cpp]: This variable must be initialized.

File: libethereum/State.h [hunk 29]
```diff
@@ -82,7 +82,8 @@ enum class BaseState
 enum class Permanence
 {
 	Reverted,
-	Committed
+	Committed,
+	Unfinished	///< uncommited state for changeLog readings
```
Comment (chfast) [inline libethereum/State.h]: In documenting comments always use full sentences. Must be `Uncommited state for change log readings in tests.".
Seeing the comment, shouldn't we change `Unfinished` to `Uncommited`?

File: libethereum/State.h [hunk 30]
```diff
@@ -142,8 +143,9 @@ struct Change
 	{}
 };
 
-}
+typedef std::vector<Change> ChangeLog;
```
Comment (chfast) [inline libethereum/State.h]: 1. `using ChangeLog = std::vector<Change>;
2. It must not be in `detail` namespace. Put it in `eth` namespace directly. Now the `detail` namespace is pointless as we now make the change log available outside of State.

File: test/tools/libtesteth/ImportTest.cpp [hunk 31]
```diff
@@ -55,10 +56,13 @@ bytes ImportTest::executeTest()
...
 	{
 		eth::Network network = eth::Network::MainNetwork;
-		std::pair<eth::State, ImportTest::execOutput> out = executeTransaction(network, m_envInfo, m_statePre, m_transaction);
-		m_statePost = out.first;
-		m_logs = out.second.second.log();
-		return out.second.first.output;
+		LogEntries emptyLogs;
+		ExecutionResult emptyRes;
+		ImportTest::ExecOutput execOut = make_pair(emptyRes, TransactionReceipt(h256(), u256(), emptyLogs));
+		std::tie(m_statePost, execOut, std::ignore) =
```
Comment (chfast) [inline test/tools/libtesteth/ImportTest.cpp]: Here this change was not necessary as ExecutionResult does not have default constructor, but it can stay.

Comment (winsvega) [inline test/tools/libtesteth/ImportTest.cpp]: we will remove this part anyway at some point

File: libethereum/State.h [hunk 32]
```diff
@@ -82,7 +82,8 @@ enum class BaseState
 enum class Permanence
 {
 	Reverted,
-	Committed
+	Committed,
+	Uncommited	///< Uncommited state for change log readings in tests.
```
Comment (chfast) [inline libethereum/State.h]: Typo, `Uncommitted`.

File: test/tools/libtesteth/FillJsonFunctions.cpp [hunk 33]
```diff
@@ -0,0 +1,218 @@
...
+	return txObject;
+}
+
+json_spirit::mObject fillJsonWithStateChange(State const& _stateOrig, eth::State const& _statePost, std::vector<dev::eth::Change> const& _changeLog)
+{
+	json_spirit::mObject oState;
+	if (!_changeLog.size())
+		return oState;
+
+	using changeKind = eth::Change::Kind;
```
Comment (chfast) [inline test/tools/libtesteth/FillJsonFunctions.cpp]: Type aliases must start from upper letter.

File: test/tools/libtesteth/FillJsonFunctions.cpp [hunk 34]
```diff
@@ -52,19 +52,19 @@ json_spirit::mObject fillJsonWithTransaction(Transaction const& _txn)
...
 {
 	json_spirit::mObject oState;
 	if (!_changeLog.size())
 		return oState;
 
-	using changeKind = eth::detail::Change::Kind;
-	using changeStruct = eth::detail::Change;
-	eth::detail::ChangeLog changeLog = _changeLog;
+	using ChangeKind = eth::Change::Kind;
+	using ChangeStruct = eth::Change;
```
Comment (chfast) [inline test/tools/libtesteth/FillJsonFunctions.cpp]: This make no sense. You can use `Change` (because `using namespace dev::eth`), why you want to replace it with `ChangeStruct`.

Similar `ChangeKind` instead of `Change::Kind`?

File: test/tools/libtesteth/FillJsonFunctions.cpp [hunk 35]
```diff
@@ -52,19 +52,19 @@ json_spirit::mObject fillJsonWithTransaction(Transaction const& _txn)
...
+	eth::ChangeLog changeLog = _changeLog;
 
 	//Sort the vector by address field
 	std::sort(changeLog.begin(), changeLog.end(),
-		[ ]( const changeStruct& lhs, const changeStruct& rhs )
+		[ ]( const ChangeStruct& lhs, const ChangeStruct& rhs )
```
Comment (chfast) [inline test/tools/libtesteth/FillJsonFunctions.cpp]: Too much spaces here. Should be: `[](const ChangeStruct& lhs, const ChangeStruct& rhs)`.

File: test/tools/libtesteth/FillJsonFunctions.cpp (hunk 34 above)
Comment (winsvega) [inline test/tools/libtesteth/FillJsonFunctions.cpp]: because we use mutation approach on this PR
